src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,196 @@
+    }
+
+    private int foundElementIndex(T element) {
+        final Object[] elements = elementData;
+        final int size = this.size;
+        int i = 0;
+        if (element == null) {
+            for (; i < size; i++) {
+                if (elements[i] == null) {
+                    return i;
+                }
+            }
+        } else {
+            for (; i < size; i++) {
+                if (element.equals(elements[i])) {
+                    return i;
+                }
+            }
+        }
+        return NOT_EXISTING_INDEX;
+    }
+
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
     }
 
     @Override
-    public int size() {
-        return 0;
+    public void add(T value) {
+        add(value, elementData, size);
     }
 
     @Override
-    public boolean isEmpty() {
-        return false;
+    public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        final int size;
+        Object[] elementData;
+        if ((size = this.size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        this.size = size + 1;
+    }
+
+    private void add(T element, Object[] elementData, int size) {
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = element;
+        this.size = size + 1;
+    }
+
+    private static int newLength(int oldLength, int minGrowth, int prefGrowth) {
+
+        int prefLength = oldLength + Math.max(minGrowth, prefGrowth);
+        if (0 < prefLength && prefLength <= SOFT_MAX_ARRAY_LENGTH) {
+            return prefLength;
+        } else {
+            return hugeLength(oldLength, minGrowth);
+        }
+    }
+
+    private static int hugeLength(int oldLength, int minGrowth) {
+        int minLength = oldLength + minGrowth;
+        if (minLength < 0) {
+            throw new OutOfMemoryError(
+                    "Required array length " + oldLength + " + " + minGrowth + " is too large");
+        } else {
+            return Math.max(minLength, SOFT_MAX_ARRAY_LENGTH);
+        }
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA) {
+            int newCapacity = newLength(oldCapacity,
+                    minCapacity - oldCapacity,
+                    oldCapacity >> 1);
+            return elementData = Arrays.copyOf(elementData, newCapacity);
+        } else {
+            return elementData = new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+        }
+    }
```

Comment:
This logic is overcomplicated actually. Try to simplify it. Your grow() method could be implemented in 3 lines actually. Also use System.arraycopy() please